### PR TITLE
http2: Add flag protected checks for frame flood and abuse by upstream servers

### PR DESCRIPTION
**Pull request by yanavlasov**
Commit Message:
Add flag protected checks for frame flood and abuse by upstream servers

Additional Description:
This is the starter PR to enable frame flood and abuse checks for upstream codecs. The check is disabled by default and can be enabled via runtime flag `envoy.reloadable_features.upstream_http2_flood_checks`. These checks will be enabled by default once all unit and integration tests are implemented.

This PR also introduces mechanism for writing arbitrary H/2 frames to the fake upstream connection.

Part of #12281

Risk Level: Low (disabled by default)
Testing: Unit Tests
Docs Changes: Yes (through proto comments)
Release Notes: Yes
Platform Specific Features: N/A
Runtime guard: envoy.reloadable_features.upstream_http2_flood_checks

Signed-off-by: Yan Avlasov <yavlasov@google.com>

### CODE REVIEW

**Review by antoniovicente — COMMENTED** [review]
Great to see that we're getting closer to having functional parity between downstream and upstream H2 flood protection.

File: test/config/utility.cc [hunk 1]
```diff
@@ -1109,6 +1109,18 @@ void ConfigHelper::setOutboundFramesLimits(uint32_t max_all_frames, uint32_t max
   }
 }
 
+void ConfigHelper::setUpstreamOutboundFramesLimits(uint32_t max_all_frames,
```
Comment (adisuissa) [inline test/config/utility.cc:1112]: Similar to @antoniovicente suggestion, can you please rename `setOutboundFramesLimits` to `setDownstreamOutboundFramesLimits`.

**Review by adisuissa — COMMENTED** [review]
Thanks for working on this!

File: docs/root/version_history/current.rst [hunk 2]
```diff
@@ -33,6 +33,7 @@ New Features
 * grpc: implemented header value syntax support when defining :ref:`initial metadata <envoy_v3_api_field_config.core.v3.GrpcService.initial_metadata>` for gRPC-based `ext_authz` :ref:`HTTP <envoy_v3_api_field_extensions.filters.http.ext_authz.v3.ExtAuthz.grpc_service>` and :ref:`network <envoy_v3_api_field_extensions.filters.network.ext_authz.v3.ExtAuthz.grpc_service>` filters, and :ref:`ratelimit <envoy_v3_api_field_config.ratelimit.v3.RateLimitServiceConfig.grpc_service>` filters.
 * hds: added support for delta updates in the :ref:`HealthCheckSpecifier <envoy_v3_api_msg_service.health.v3.HealthCheckSpecifier>`, making only the Endpoints and Health Checkers that changed be reconstructed on receiving a new message, rather than the entire HDS.
 * health_check: added option to use :ref:`no_traffic_healthy_interval <envoy_v3_api_field_config.core.v3.HealthCheck.no_traffic_healthy_interval>` which allows a different no traffic interval when the host is healthy.
+* http: added frame flood and abuse checks to the upstream HTTP/2 codec. This check is off by default and can be enabled by setting the `envoy.reloadable_features.upstream_http2_flood_checks` runtime key to true.
```
[the remark this answers is not in the document]
Comment (yanavlasov) [reply docs/root/version_history/current.rst:36]: done

File: source/common/http/codec_client.cc [hunk 3]
```diff
@@ -140,6 +140,12 @@ void CodecClient::onData(Buffer::Instance& data) {
     if (!active_requests_.empty() || getPrematureResponseHttpCode(status) != Code::RequestTimeout) {
       protocol_error = true;
     }
+  } else if (isBufferFloodError(status) || isInboundFramesWithEmptyPayloadError(status)) {
+    ENVOY_CONN_LOG(debug, "flood detected: {}", *connection_, status.message());
+    close();
+    protocol_error = true;
```
[the remark this answers is not in the document]
Comment (yanavlasov) [reply source/common/http/codec_client.cc]: done

File: source/common/http/http2/codec_impl.h [hunk 4]
```diff
@@ -563,17 +563,9 @@ class ClientConnectionImpl : public ClientConnection, public ConnectionImpl {
...
-  // the clean-up of upstream connections. The clean-up of upstream connection causes RST_STREAM
-  // messages to be sent on corresponding downstream connections. This may actually trigger flood
-  // mitigation on the downstream connections, however there is currently no mechanism for
-  // handling these types of errors.
-  // TODO(yanavlasov): add flood mitigation for upstream connections as well.
-  ProtocolConstraints::ReleasorProc trackOutboundFrames(bool) override {
-    return ProtocolConstraints::ReleasorProc([]() {});
-  }
-  Status trackInboundFrames(const nghttp2_frame_hd*, uint32_t) override { return okStatus(); }
+  // TODO(yanavlasov): move to the base class once upstream flood checks are not flag protected.
```
[the remark this answers is not in the document]
Comment (yanavlasov) [reply source/common/http/http2/codec_impl.h]: done

File: test/config/utility.h [hunk 5]
```diff
@@ -223,9 +223,12 @@ class ConfigHelper {
   // and write it to the lds file.
   void setLds(absl::string_view version_info);
 
-  // Set limits on pending outbound frames.
+  // Set limits on pending downstream outbound frames.
   void setOutboundFramesLimits(uint32_t max_all_frames, uint32_t max_control_frames);
```
[the remark this answers is not in the document]
Comment (yanavlasov) [reply test/config/utility.h]: done

File: test/integration/fake_upstream.h [hunk 6]
```diff
@@ -614,6 +614,13 @@ class FakeUpstream : Logger::Loggable<Logger::Id::testing>,
     return Http::Http2::CodecStats::atomicGet(http2_codec_stats_, stats_store_);
   }
 
+  // Write into the outbound buffer of the network connection at the specified index.
+  // Note: that this write bypasses any processing by the upstream codec.
+  ABSL_MUST_USE_RESULT
+  testing::AssertionResult
+  writeConnection(uint32_t index, const std::string& data, bool end_stream = false,
```
[the remark this answers is not in the document]
Comment (yanavlasov) [reply test/integration/fake_upstream.h]: done

File: test/config/utility.cc (hunk 1 above)
Comment (yanavlasov) [reply test/config/utility.cc:1113]: done

**Review by antoniovicente — APPROVED** [review]
Thanks!

**Pull request merged**